src/credential_provider.rs

```diff
@@ -0,0 +1,432 @@
+                        source: Some(crate::ErrorSource::Unknown(Box::new(e))),
+                    })
+                }
+            }
+        } else if let Some(auth_token) = self.auth_token_string {
+            if auth_token.is_empty() {
+                return Err(MomentoError::InvalidArgument {
+                    description: "Auth token string cannot be empty".into(),
+                    source: None,
+                });
+            }
+            auth_token
+        } else {
+            return Err(MomentoError::InvalidArgument {
+                description: "Either environment variable or auth token string must be provided"
+                    .into(),
+                source: None,
+            });
+        };
+
+        CredentialProviderBuilder::decode_auth_token(
+            token_to_process,
+            self.cache_endpoint_override,
+            self.control_endpoint_override,
+        )
+    }
+
+    fn decode_auth_token(
+        auth_token: String,
+        cache_endpoint_override: Option<String>,
+        control_endpoint_override: Option<String>,
+    ) -> MomentoResult<CredentialProvider> {
+        match base64::engine::general_purpose::URL_SAFE.decode(&auth_token) {
+            Ok(auth_token_bytes) => CredentialProviderBuilder::process_v1_token(
+                auth_token_bytes,
+                cache_endpoint_override,
+                control_endpoint_override,
+            ),
+            Err(_) => CredentialProviderBuilder::process_jwt_token(
+                auth_token,
+                cache_endpoint_override,
+                control_endpoint_override,
+            ),
+        }
+    }
+
+    fn process_v1_token(
+        auth_token_bytes: Vec<u8>,
+        cache_endpoint_override: Option<String>,
+        control_endpoint_override: Option<String>,
+    ) -> MomentoResult<CredentialProvider> {
+        let json: V1Token = serde_json::from_slice(&auth_token_bytes)
+            .map_err(|e| token_parsing_error(Box::new(e)))?;
+
+        // If endpoint override is present then that always takes precedence over the
+        // endpoint from the token
+        let cache_endpoint = cache_endpoint_override.unwrap_or(
+            CredentialProviderBuilder::get_cache_endpoint(json.endpoint.clone()),
+        );
+        let control_endpoint = control_endpoint_override.unwrap_or(
+            CredentialProviderBuilder::get_control_endpoint(json.endpoint),
+        );
+
+        Ok(CredentialProvider {
+            auth_token: json.api_key,
+            cache_endpoint: CredentialProviderBuilder::https_endpoint(cache_endpoint),
+            control_endpoint: CredentialProviderBuilder::https_endpoint(control_endpoint),
+        })
+    }
+
+    fn process_jwt_token(
+        auth_token: String,
+        cache_endpoint_override: Option<String>,
+        control_endpoint_override: Option<String>,
+    ) -> MomentoResult<CredentialProvider> {
+        let key = DecodingKey::from_secret("".as_ref());
```

Comment:
Why are you as_reffing a ref here? Isn't `""` enough of a ref? It's an `&str`, and more specifically a `&'static str`

Reply:
Because I copied the code over from another file where it was like this 😅 .

Reply:
Ah looks like it's actually because `DecodingKey::from_secret()` takes a `&[u8]` not `&str`. `as_ref()` under the hood does a `as_bytes` ... I'm gonna do `b""`

Reply:
Great!

Reply:
Fixed in [ed9fe51](https://github.com/momentohq/client-sdk-rust/pull/146/commits/ed9fe519693969269ef07187f90f851b68a9e689)